crates/oxc_linter/src/rules/unicorn/prefer_dom_node_dataset.rs

```diff
@@ -0,0 +1,213 @@
+use oxc_ast::{
+    ast::{Argument, Expression},
+    AstKind,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::{self, Error},
+};
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint-plugin-unicorn(prefer-dom-node-dataset): Prefer using `dataset` over `{1}`")]
+#[diagnostic(severity(warning))]
+struct PreferDomNodeDatasetDiagnostic(#[label] pub Span, pub String);
+
+#[derive(Debug, Default, Clone)]
+pub struct PreferDomNodeDataset;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Use [`.dataset`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset) on DOM elements over `getAttribute(…)`, `.setAttribute(…)`, `.removeAttribute(…)` and `.hasAttribute(…)`.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// The `dataset` property is a map of strings that contains all the `data-*` attributes from the element. It is a convenient way to access all of them at once.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// // Bad
+    /// element.setAttribute('data-unicorn', '🦄');
+    ///
+    /// // Good
+    /// element.dataset.unicorn = '🦄';
+    /// ```
+    PreferDomNodeDataset,
+    pedantic
+);
+
+impl Rule for PreferDomNodeDataset {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        let Some(member_expr) = call_expr.callee.get_member_expr() else { return };
+
+        if member_expr.is_computed() {
+            return;
+        }
+
+        let Some(method_name) = member_expr.static_property_name() else { return };
```

Comment:
It seems like we have a lot of these API checks, we can have a follow up PR where we refactor all these into a common util.

Reply:
yeah i've noticed that also